Type label support for pair_style commands

jrgissing opened the pull request:
**Summary**

Add type label support for certain pair_style commands that have atom, bond, or angle type arguments

- [x] lj/cut/tip4p/long (KSPACE)
  - [x] corrected*
- [x] lj/long/tip4p/long (KSPACE)
  - [x] corrected*
- [x] tip4p/long (KSPACE)
  - [x] corrected*
- [x] tip4p/long/soft (FEP)
  - [x] corrected*
- [x] lj/cut/tip4p/long/soft (FEP)
  - [x] corrected*
- [x] tip4p_cut (MOLECULE)
  - [x] corrected*
- [x] pair_lj_cut_tip4p_cut (MOLECULE)
  - [x] corrected*
- [ ] eff - skip (essentially an element list, similar to other pair styles)
- [x] pair_coeff for pair_style atm
- [x] pair_coeff for hbond/dreiding/lj
- [x] pair_coeff for hbond/dreiding/morse
- [x] e3b
  - [x] corrected*
- [x] srp
  - [x] corrected*
- [x] pair_write
- [x] add checks for legal atom types
- [ ] Update TIP4P HOWTO with type labels?
- [ ] others? (above found by 'grep "=.\*atom type" pair_\*' in docs/src)

*refactored to not evaluate type labels in pair_style command immediately, but instead when coeff() is called


**Related Issue(s)**

none

~~OPT, GPU, OPENMP packages not checked~~ (should be fine, settings not read in, in these routines)

**Author(s)**

Jake

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

jrgissing committed:
Update pair_lj_cut_tip4p_long.cpp

jrgissing committed:
type labels for other KSPACE pair_styles

jrgissing committed:
docs for tip4p type labels update

lj/cut and lj/long

jrgissing committed:
type labels for FEP tip4p styles

jrgissing committed:
copy/paste error

jrgissing committed:
docs for FEP tip4p type labels

jrgissing committed:
type labels for MOLECULE tip4p styles

jrgissing committed:
wrong link

jrgissing committed:
Update pair_coul.rst

jrgissing committed:
type label for pair_coeff atm K value

jrgissing committed:
type labels for hbond pair styles K value

jrgissing committed:
type label for pair e3b oxygen

jrgissing committed:
type labels for pair_srp bond type

jrgissing committed:
type label support for pair_write

jrgissing commented:
I am realizing that pair_style is often called before types or type labels are defined. In the case of tip4p styles, for example, there is no check for whether typeO is a valid atom type. Perhaps these type of checks can be added in a routine that is called later than settings(), such as ~~init_style()~~ coeff()

jrgissing committed:
proposed solution

bit of trickery to allow type labels to be used before they are defined

jrgissing committed:
refactor lj/long/tip4p/long

jrgissing committed:
refactor tip4p/long

jrgissing committed:
convert rest of tip4p

jrgissing committed:
convert e3b and srp

akohlmey committed:
must not set tip4p atom and bond types when restarting from a restart file

akohlmey commented:
@jrgissing I figured it out. It is much less severe than I suspected. When calling `Pair::coeff()` after a restart the type strings are empty and the corresponding variables for types already set. You can see the error when running:
```
TEST_ARGS=-v ctest -R tip4p -V
```
This will not only show the test results but also echo the commands and their output. 

Changes are in commit 5da737fd11de324e891099c3c3b551c8e22c32a7

akohlmey committed:
Merge branch 'develop' into type-labels-for-pair_style_commands

jrgissing commented:
Thank you! I suspect I will need to port that fix over to some of the non-TIP4P styles, even if they are not triggering an unit test failure. I should be able to do that later tonight

jrgissing committed:
Update pair_e3b.cpp

jrgissing committed:
Update pair_srp.cpp

akohlmey committed:
fix issue with pair style srp that broke regression test

jrgissing committed:
expand_type_int: optional flag to check types

verifying type could probably be default, except for historical reasons
would involve removing redundant checks after calling routine

jrgissing committed:
expand_type_int verify->bool
